server/src/main/java/org/elasticsearch/node/ResponseCollectorService.java

```diff
@@ -58,25 +58,23 @@ void removeNode(String nodeId) {
     }
 
     public void addNodeStatistics(String nodeId, int queueSize, long responseTimeNanos, long avgServiceTimeNanos) {
-        nodeIdToStats.compute(nodeId, (id, ns) -> {
-            if (ns == null) {
-                ExponentiallyWeightedMovingAverage queueEWMA = new ExponentiallyWeightedMovingAverage(ALPHA, queueSize);
-                ExponentiallyWeightedMovingAverage responseEWMA = new ExponentiallyWeightedMovingAverage(ALPHA, responseTimeNanos);
-                return new NodeStatistics(nodeId, queueEWMA, responseEWMA, avgServiceTimeNanos);
-            } else {
-                ns.queueSize.addValue((double) queueSize);
-                ns.responseTime.addValue((double) responseTimeNanos);
-                ns.serviceTime = avgServiceTimeNanos;
-                return ns;
-            }
-        });
+        var ns = nodeIdToStats.get(nodeId);
+        if (ns == null) {
+            ns = new NodeStatistics(nodeId, queueSize, responseTimeNanos, avgServiceTimeNanos);
+            ns = nodeIdToStats.putIfAbsent(nodeId, ns);
+        }
+        if (ns != null) {
+            ns.serviceTime = avgServiceTimeNanos;
```

Comment:
This assignment is kinda racy but moving it before the volatile writes in the following lines probably fixes things in practice because we only read the variable after observing volatile writes on the other 2 fields now.